SQL Translation for DateTimeOffset members

JakenVeina opened the pull request:
Added SQL Translation support for members of DateTimeOffset values, such as static member DateTimeOffset.Now, and instance members .Date, .Year, .AddMinutes(), etc.

Relates to (and possibly resolves) #473.


So, after having actually implemented and tested this, I feel I really have to mention... I'm not convinced this is something EFCore.PG SHOULD support. There's some value, since `DateTimeOffset` is a supported type mapping, but there's a couple of very serious pitfalls. Well, I suppose just one pitfall, but a variety of ways it could manifest.

The simplest operation that can manifest this is something like...
```sql
SELECT "Timestamp" AT TIME ZONE 'UTC'
FROM "TestRecords"
```
...where `"Timestamp"` is a `TIMESTAMP WITH TIME ZONE` column. In this case, PostgreSQL handles the conversion, just fine. But as defined [here](https://www.npgsql.org/doc/types/datetime.html), `Npgsql` really doesn't support reading `DateTimeOffset` into .NET. A `TIMESTAMP WITH TIME ZONE` value will be translated into an `NpgsqlDateTime` value, in the client-local timezone, with correct offset adjustments, but at this point the offset info is dropped and cannot be recovered.

In an application using `Npgsql` directly, code could be written in-context to know that the `DateTime` values should be interpreted as `'UTC'` and a correct `DateTimeOffset` could be constructed, if so desired. However, EFCore.PG has no such context, and can only perform a translation with the `NpgsqlDateTime` object. Currently, EFCore.PG just assumes the `DateTime` value is for the local timezone, which is incorrect in this case.

Long story short, an expression of `x.MyDateTimeOffset.ToUniversalTime()` evaluated on the database versus having `.ToUniversalTime()` evaluated on the client will always result in mismatched values (unless the database's timezone is `'UTC'`).

I still included translations for `.ToUniversalTime()` because I think the other member translations are almost useless without it. I.E. using `x.MyDateTimeOffset.Date` in a database-evaluated query can result in unpredictable or unexpected results, since you might not be sure of the offset of all values involved, but `x.MyDateTimeOffset.ToUniversalTime().Date` can be quite useful. This is exactly the use-case that started me on this whole trek, as we wanted an analytic query that does a `.GroupBy()` over a `.Date`.

In a perfect world, I think I'd say that support for `DateTimeOffset` in EFCore.PG should be removed, unless a new type is added to `Npgsql` to support it, without dropping the offset. Of course, that would be a heavily breaking change, so probably never gonna happen.

If anyone does think this has enough value to be pulled, I realize I haven't added any testing for these changes. I actually have written some tests in a separate project, that I didn't include here, because I didn't even know where to start to get them integrated into the existing test projects. I couldn't even get them to run. It seems like they're trying to contact a test database, to which I don't have access? Or, I just need to setup a local PostgreSQL instance? Regardless, I'm happy to add tests to the PR if anyone can give me guidance on how to do so.

roji commented:
@JakenVeina sorry for not commenting/reviewing here earlier. I see that you've merged in the latest main, so it seems like you're still interested in getting this merge, right?

> Npgsql really doesn't support reading DateTimeOffset into .NET. A TIMESTAMP WITH TIME ZONE value will be translated into an NpgsqlDateTime value, in the client-local timezone, with correct offset adjustments, but at this point the offset info is dropped and cannot be recovered.

I don't really follow this... Npgsql does support mapping DateTimeOffset to PG timestamptz - including at the EF Core level. In fact, NpgsqlDateTime is the one that isn't supported at the EF Core level (it has to be explicitly requested at the ADO level). Can you also elaborate on what can be done better with NpgsqlDateTime with Npgsql?

My general feeling about 

(note that the offset info isn't really dropped since it never existed in the first place - timestamptz does not store it. All this type does in PG is assume that UTC is being stored, and converted that on output/input to the user's local timezone based on the connection's Timezone parameter. Now, since Npgsql uses binary encoding, PostgreSQL doesn't perform these timezone conversions, and so Npgsql mimics the same behavior when reading/writing the type - to the local machine's configured time zone rather than to the PG connection Timezone. But all that probably doesn't matter much.)

> Long story short, an expression of x.MyDateTimeOffset.ToUniversalTime() evaluated on the database versus having .ToUniversalTime() evaluated on the client will always result in mismatched values (unless the database's timezone is 'UTC').

On of the main usages of translating anything is allowing users to express it in a LINQ Where predicate. I'm not following what would be bad about letting users write:

```c#
var blogs = ctx.Blogs.Where(b => b.Creation.ToUniversalTime() > something).ToList();
```

and translating that for server evaluation...

PS This would indeed need to add some testing. There are various EF Core tests which are currently skipped for exactly these properties.
PPS .NET has DateTime.Today - which we translate - but DateTimeOffset doesn't seem to, so we should exclude that.

JakenVeina commented:
> sorry for not commenting/reviewing here earlier. I see that you've merged in the latest main, so it seems like you're still interested in getting this merge, right?

Not a problem. I'd honestly forgotten about it until I spotted it in my PRs list and saw that there were conflicts. I'm also gonna have to try and re-learn what this problem was as I go through this, so bear with me. ;)

The original original original problem, that we encountered was basically this (paraphrasing):

```cs
var messageCountsByDate = await context.Messages
    .Where(m => m.Timestamp > (DateTimeOffset.Now - TimeSpan.FromDays(30)))
    .GroupBy(
        m => m.Timestamp.Date,
        g => g.Count())
    .ToDictionaryAsync();
```

This could not be translated, because `m.Timestamp` is `DateTimeOffset` and the Npgsql translator does not translate `DateTimeOffset.Date`. In the process of writing this, and writing tests, I realized that allowing for translation of `DateTimeOffset` values actually allows for some pitfalls (I.E. incorrect query results), resulting from the fact that Npgsql doesn't support a type of its own that parallels `DateTimeOffset`.

>  Npgsql does support mapping DateTimeOffset to PG timestamptz.

IIRC:

The issue is in the reverse direction. When a `TIMESTAMP WITH TIME ZONE` value is being read into an `NpgsqlDateTime` value, Somewhere along the trip, a conversion happens that results in the final `NgpsqlDateTime` value being localized to the .NET client, regardless of what was asked for in SQL. And that's not a HUGE deal when working with `NpgsqlDateTime` directly, since the data in that struct is always self-consistent. But when EF tries to go an extra step and convert from `NpgsqlDateTime` to `DateTimeOffset` it can get it wrong, because it doesn't have enough info.

> I'm not following what would be bad about letting users write: `var blogs = ctx.Blogs.Where(b => b.Creation.ToUniversalTime() something).ToList();` and translating that for server evaluation...

That's exactly what worries me. That SEEMS like a perfectly reasonable query that the user should be able to perform, and have translate. But it can result in an incorrect data conversion. I'll have to pull up some code again and try and replicate this.

> Can you also elaborate on what can be done better with NpgsqlDateTime with Npgsql?

I'll reserve my comment until I replicate the pitfall again, but I recall that I didn't think there was anything wrong with `NpgsqlDateTime` itself.

> PS This would indeed need to add some testing. There are various EF Core tests which are currently skipped for exactly these properties.

Looks like the tests I had written are gone. I wiped out my local clone at one point, and I must've not had them checked in. Should be able to re-write them though, would just need some guidance on where they belong in the architecture, and how to get the test project to run.

> PPS .NET has DateTime.Today - which we translate - but DateTimeOffset doesn't seem to, so we should exclude that.

That should be accounted for in the PR.

roji commented:
>> I'm not following what would be bad about letting users write: var blogs = ctx.Blogs.Where(b => b.Creation.ToUniversalTime() something).ToList(); and translating that for server evaluation...
>
> That's exactly what worries me. That SEEMS like a perfectly reasonable query that the user should be able to perform, and have translate. But it can result in an incorrect data conversion.

The important thing to understand, is that in the query above timestamptz isn't even being transferred to the client, or converted to NpgsqlDateTime or DateTimeOffset, or anything else; it's in a WHERE clause that's evaluated purely on the server and results in a boolean. So the translation of ToUniversalTime serves purely for the user to express some SQL that gets sent to the server.

Re the specific conversion problems that you mentioned above - which would be relevant if the timestamptz *is* being transferred to the client - it would be good to see a very concrete scenario where you thing the end results are wrong, we can discuss on that.

JakenVeina commented:
> The important thing to understand, is that in the query above timestamptz isn't even being transferred to the client, or converted to NpgsqlDateTime or DateTimeOffset, or anything else; it's in a WHERE clause that's evaluated purely on the server and results in a boolean. So the translation of ToUniversalTime serves purely for the user to express some SQL that gets sent to the server.

Right, in that situation, everything works great, with the changes I implemented. In a situation where the caller actually DOES want to transmit all that info up to .NET, it comes out incorrect.

Really, I think what we're talking about is a bug or design defect that's been around for a long time, but never manifested until I added these new query translations for `DateTimeOffset`.

Specific example:

```cs
var result1 = await context.Set<TestEntity>()
    .AsQueryable()
    .Where(t => t.Id == 1)
    .Select(t => t.Created.ToUniversalTime())
    .FirstAsync();

var result2 = await context.Set<TestEntity>()
    .AsQueryable()
    .Where(t => t.Id == 1)
    .AsAsyncEnumerable()
    .Select(t => t.Created.ToUniversalTime())
    .FirstAsync();

Console.WriteLine(result1);
Console.WriteLine(result2);
Console.WriteLine(result1 == result2);
```

The only difference between these two queries is whether `.ToUniversalTime()` runs server-side or client-side. But the end result is that `result1` and `result2` do not match.

```
2/2/2001 9:05:06 PM -06:00
2/2/2001 9:05:06 PM +00:00
False
```

The compiled SQL for the `result1` query is:

```sql
SELECT t."Created" AT TIME ZONE 'UTC'
FROM "TestEntity" AS t
WHERE t."Id" = 1
LIMIT 1
```

Looking at the database directly, I can see that the value stored is `2001-02-02 15:05:06-06`, meaing that `result2` where `.ToUniversalTime()` was calculated client-side, is correct. It also matches the value `2001-02-03 04:05:06 +7:00` from the original insert of that record, meaning all the conversions within the client-to-database path are correct.


If I take the above query and execute it directly with an `NpgsqlCommand` I get a `DateTime` value of `2/2/2001 9:05:06 PM`, which is correct, EXCEPT that the `.Kind` value is `Unspecified`. This is probably the core reason why EF incorrectly converts this value to `DateTimeOffset`. EF seems to simply assume this value is localized, when it isn't. 

I feel like the only way to TRULY deal with this problem would be to enhance `Npgsql` to include a new `NgpsqlDateTimeOffset` type, or something similar, which includes the transmission of offset data directly from the database. I.E. when the user explicitly does a timezone conversion in the database, the only way to accurately transmit that value up to the client is to include the offset value in the data transmission.

roji commented:
> I feel like the only way to TRULY deal with this problem would be to enhance Npgsql to include a new NgpsqlDateTimeOffset type, or something similar, which includes the transmission of offset data directly from the database. I.E. when the user explicitly does a timezone conversion in the database, the only way to accurately transmit that value up to the client is to include the offset value in the data transmission.

The point is, there is no offset data for us to transmit from the database... Regardless of whether the user explicitly does a timezone conversion in the database or not, PG doesn't send *any* timezone information with the timestamp, ever. That's just how the timestamptz type works in PostgreSQL.

> This is probably the core reason why EF incorrectly converts this value to DateTimeOffset. EF seems to simply assume this value is localized, when it isn't.

I dug a bit into this, you're not far off the mark. Here's what's going on:

* You initially insert '2001-02-03 04:05:06+7:00'::timestamptz into the database. After conversion to UTC that's 2001-02-02 21:05:06.000000 in the database (remember, no actual offset in the database).
* In your EF query, ToUniversalTime gets translated to PG `AT TIME ZONE`. That expression changes the PG type from `timestamp with time zone` to `timestamp without time zone` ([see PG docs](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-ZONECONVERT)). We do this accordingly in our query pipeline (see [NpgsqlSqlExpressionFactory.FlipTimestampTypeMapping](https://github.com/npgsql/efcore.pg/blob/main/src/EFCore.PG/Query/NpgsqlSqlExpressionFactory.cs#L65)).
* Since you're selecting this value out, and PG `timestamp without time zone` is mapped to BCL DateTime, EF reads a DateTime from NpgsqlDataReader. Npgsql properly returns a DateTime with 2001-02-02 21:05:06.000000, Kind=unspecified. Up to here all is more or less well, the interesting part comes next.
* The CLR type coming back from the database (DateTime) doesn't correspond to the CLR type returned by the Select (DateTimeOffset), which is what EF is supposed to return. So the EF shaper casts the DateTime to DateTimeOffset.
* In .NET, casting a DateTime with Kind=Unspecified to a DateTimeOffset simply takes the DateTime fields as-is, and adds the local machine's timezone. This is how you get the values from the database, but with the local timezone tacked on. This may be questionable behavior, but it is what it is.

Stepping back, I agree that the above makes ToUniversalTime useless/incorrect when used in projections out of the database. Thinking about it some more, it also seems useless in predicates: since `timestamp with time zone` values are already supposed to be in UTC in the database, the only effect of `AT TIME ZONE` here would be to flip the type mapping to `timestamp without time zone`, which doesn't seem useful. So I'm fine with not translating this particular functionality.

This made me think about our current translation of ToUniversalTime for DateTime, and I'm not sure it makes sense either. In projections, `AT TIME ZONE` flips `timestamp` to `timestamptz`; this makes Npgsql convert the value to a DateTime with Kind=Local, which is the opposite of what was originally requested (the user called ToUniversalTime). In predicates, as with DateTimeOffset above, all the `AT TIME ZONE 'UTC'` does is flip the PG type from `timestamp` to `timestamptz`. That might be useful, I'm not sure - it's basically a PG type cast.

Finally... We've been spending a lot of time on ToUniversalTime - but that's the less important member on DateTimeOffset :) I'd indeed be fine with not translating ToUniversalTime, but we can get the rest in.

JakenVeina commented:
> That expression changes the PG type from timestamp with time zone to timestamp without time zone

That'll do it. Makes sense in the context of Npgsql, because you don't NEED a timezone to be attached. The caller is assumed to know what the timezone is because they explicitly asked for it. When it all runs through layers of EF translation, it's kinda impossible for EF to know. Unless there's some way to attach the intended timezone as metadata to a top-level column of a query, so the converter could extract it and use it during conversion?

> I'd indeed be fine with not translating ToUniversalTime, but we can get the rest in.

The other members arguably become a bit less useful, if `.ToUniversalTime()` isn't there as well. For example, the use-case I described earlier. I left out a `.ToUniversalTime()` call there.

```cs
var messageCountsByDate = await context.Messages
    .Where(m => m.Timestamp > (DateTimeOffset.Now - TimeSpan.FromDays(30)))
    .GroupBy(
        m => m.Timestamp.ToUniversalTime().Date,
        g => g.Count())
    .ToDictionaryAsync();
```

Maybe there's some way to detect when `.ToUniversalTime()` is being used at the top level, and throw an exception to prevent it?

roji commented:
> That'll do it. Makes sense in the context of Npgsql, because you don't NEED a timezone to be attached. The caller is assumed to know what the timezone is because they explicitly asked for it. When it all runs through layers of EF translation, it's kinda impossible for EF to know. Unless there's some way to attach the intended timezone as metadata to a top-level column of a query, so the converter could extract it and use it during conversion?

Not really. At the end of the day, PG sends us data back, Npgsql reads those values at the ADO level and the EF provider accesses them. You seem to be thinking of some way to do something in SQL that would affect what PostgreSQL sends us back, in a way that allows us to know what the intended timezone is - I'm not aware of anything like this. Again, PostgreSQL does not have a timestamp type that carries timezone information on it.

> The other members arguably become a bit less useful, if .ToUniversalTime() isn't there as well. For example, the use-case I described earlier. I left out a .ToUniversalTime() call there.

I wouldn't say that's true... There are various scenarios where someone simply wants to access some property on DateTimeOffset that's mapped to a timestamptz, e.g. in some WHERE clause. That doesn't seem to depend on ToUniversalTime in any way.

> The other members arguably become a bit less useful, if .ToUniversalTime() isn't there as well. For example, the use-case I described earlier. I left out a .ToUniversalTime() call there.

Can you explain what you expect ToUniversalTime to do in your GroupBy clause above, and what the type of m.Timestamp is? 

> Maybe there's some way to detect when .ToUniversalTime() is being used at the top level, and throw an exception to prevent it?

Something like that may be doable, but not trivial. I'm still not understand what value you're finding with ToUniversalTime that justifies translating it at all.

JakenVeina commented:
> You seem to be thinking of some way to do something in SQL that would affect what PostgreSQL sends us back, in a way that allows us to know what the intended timezone is - I'm not aware of anything like this. Again, PostgreSQL does not have a timestamp type that carries timezone information on it.

Right. PostgreSQL doesn't support this. It COULD support this, in theory. Other RDBMS systems do. Oracle *shudder*, for example, does store actual timezone offsets in a native datatype, and propagates that value all the way up to .NET.

The bit about "metadata" was just me musing about the EF translation layer, about having some way to implement a `DateTime` to `DateTimeOffset` converter that doesn't just do the conversion with `DateTimeKind.Unspecified` but refers back to analysis performed during the query compilation step to figure out if a timezone was specified within the SELECT expression for the column being converted.

> Can you explain what you expect ToUniversalTime to do in your GroupBy clause above, and what the type of m.Timestamp is?

`m.Timestamp` is a `DateTimeOffset` in the EF entity model. `m.Timestamp.ToUniversalTime().Date` I should think is rather self-explanatory. It's doing a timezone conversion BEFORE truncating the time portion of the timestamp, because the timezone that the timestamp is currently expressed in affects the output of that truncation. `2021-01-02 02:00:00 +0:00" and `2021-01-01 20:00:00 -6:00" are "equal" timestamps, but truncating the time portion to get just the date produces different values.

For that analytic message-count query, the IDEAL way to do it would be to either...
A) convert all the timestamps to the timezone of the user that requested the data, so they're seeing the counts from the perspective of their timezone, or
B) convert all the timestamps to a common timezone, like UTC, so all users see the same results.

I admit it's a pretty niche use-case, and it's not worth supporting if it introduces the possibility for incorrect data to be returned through EF, and there's no EF extension points to work around that.

roji commented:
> Right. PostgreSQL doesn't support this. It COULD support this, in theory. Other RDBMS systems do. Oracle shudder, for example, does store actual timezone offsets in a native datatype, and propagates that value all the way up to .NET, when queried.

Yep. SQL Server has a DateTimeOffset - though an offset is not a timezone (daylight savings can bite you hard). PG does have timetz which has an offset, but no timestamp.

> The bit about "metadata" was just me musing about the EF translation layer, about having some way to implement a DateTime to DateTimeOffset converter that doesn't just do the conversion with DateTimeKind.Unspecified but refers back to analysis performed during the query compilation step to figure out if a timezone was specified within the SELECT expression for the column being converted.

Uh... not really....

>> Can you explain what you expect ToUniversalTime to do in your GroupBy clause above, and what the type of m.Timestamp is?
>
> It's doing a timezone conversion BEFORE truncating the time portion of the timestamp, because the timezone that the timestamp is currently expressed in affects the output of that truncation.

A timezone conversion to what? I think you're still missing the important point that there *is* no offset in the database, and that m.Timestamp is already UTC, since it's a timestamptz column. Since your expression - `m.Timestamp.ToUniversalTime().Date` - is fully evaluated inside the database, ToUniversalTime simply isn't needed (and only does harm). Simply doing `m.Timestamp.Date` on a timestamptz column gets the UTC date for your timestamptz.

I know it's difficult - there's a fundamental mismatch between the .NET types (and sometimes even methods) and the PG world. DateTimeOffset does *not* really correspond to timestamptz; but we use it as a sort of approximation, since it's better than nothing at all. You may want to look at NodaTime and our support for that - things are much better/clearer there.

>  A) convert all the timestamps to the timezone of the user that requested the data, so they're seeing the counts from the perspective of their timezone, or
> B) convert all the timestamps to a common timezone, like UTC, so all users see the same results. Of course, this is what happens without any timezone conversion, it's just that the common timezone is whatever the .NET server's timezone is.

Once again, to get the right counts, there's no need to convert anything, since the timestamps have no timezone in the database. It's just timestamptz values, which are pure timestamps that are *assumed* to be UTC.

JakenVeina commented:
> A timezone conversion to what?
To what the consumer requested. Because the user can *ask* PG to perform a timezone conversion (not necessarily just to UTC), because maybe the user doesn't *want* all the timestamps to be in UTC before the time portion is truncated. Because the result of truncating time is *different* depending on what timezone the timestamp is "assumed" to be in. I described an example where we want to count messages "from the perspective" of the timezone of the end user, meaning that different users with different timezones would want to see different counts. Maybe this particular scenario isn't something anyone would ever want, but it doesn't strike me as fundamentally invalid.

> I think you're still missing the important point that there is no offset in the database
I promise, I do get that. A column of type TIMESTAMP WITH TIME ZONE is stored physically as a plain timestamp, in UTC. No offset is stored.

> I know it's difficult
Story of dates and times in programming.

roji commented:
> maybe the user doesn't want all the timestamps to be in UTC before the time portion is truncated

But that's the definition of the timestamptz type. It contains timestamps assumed to be UTC. That's why I'm not understanding why the user would want to ask PostgreSQL to convert them to... UTC (I'm talking specifically about ToUniversalTime here). There is no user perspective or assumptions here - timestamptz is UTC, so the DateTimeOffset property you're using to express the timestamp column is always UTC as well.

Think of it this way - when it's mapped to PostgreSQL by the EF Core provider, DateTimeOffset always has its offset set to UTC. Always.

Now, regarding *other* timezone transformations - *not* to UTC... We could translate [DateTimeOffset.ToOffset](https://docs.microsoft.com/en-us/dotnet/api/system.datetimeoffset.tooffset?view=net-5.0#System_DateTimeOffset_ToOffset_System_TimeSpan_) to PostgreSQL `AT TIME ZONE xxx`. It seems like it sensible - it would allow users to express, in a LINQ query, that PG should take a timestamptz (UTC!) and convert it to some arbitrary offset (e.g. so that they can extract a date out of it, like you do above). However, on the PG side, the result of `AT TIME ZONE` applied to timestamptz is a timestamp, whereas on the .NET side ToOffset returns a DateTimeOffset... So at that point things become quite confused as well.

JakenVeina commented:
> Now, regarding other timezone transformations - not to UTC... We could translate DateTimeOffset.ToOffset to PostgreSQL AT TIME ZONE xxx. It seems like it sensible - it would allow users to express, in a LINQ query, that PG should take a timestamptz (UTC!) and convert it to some arbitrary offset (e.g. so that they can extract a date out of it, like you do above). However, on the PG side, the result of AT TIME ZONE applied to timestamptz is a timestamp, whereas on the .NET side ToOffset returns a DateTimeOffset... So at that point things become quite confused as well.

Yeah, it wouldn't make much sense to translate `.ToUniversalTime()` without also doing `.ToOffset()`, and I think we've agreed we shouldn't add `.ToUniversalTime()` without a way to mitigate the `DateTimeKind.Unspecified` issue, even if it has legitimate use-cases.

roji commented:
> Yeah, it wouldn't make much sense to translate .ToUniversalTime() without also doing .ToOffset(), and I think we've agreed we shouldn't add .ToUniversalTime() without a way to mitigate the DateTimeKind.Unspecified issue, even if it has legitimate use-cases.

IMHO neither make any sense - either in isolation or together. Things just don't map well from .NET to PostgreSQL here.

However, I still think translating the other members is useful; it would allow users to write:

```c#
var blogs = context.Blogs.Where(t => t.CreationTime.Date == ...).ToList();
```

... extracting blogs created on a specific UTC date (again, assuming that CreationTime is a DateTimeOffset, and therefore mapped to timestamptz which is always assumed to be UTC). Then again, I don't see the value of using DateTimeOffset/timestamptz vs. just using DateTime/timestamp, which is why I've never bothered to translate any members on it.

roji commented:
In version 6.0 we've done some very major changes to how timestamps are handled (see [this blog post](https://www.roji.org/postgresql-dotnet-timestamp-mapping) and the [6.0 release notes](https://www.npgsql.org/efcore/release-notes/6.0.html?tabs=annotations)). Some of what I've written above is no longer relevant, and a lot of translations for DateTimeOffset members/methods have been added as they make much more sense in the new timestamp mapping - including e.g. ToUniversalTime ([see the translation docs](https://www.npgsql.org/efcore/mapping/translations.html#date-and-time-functions)).

So I'm going to go ahead and close this PR. If any DateTimeOffset translations are missing, please open a separate issue for that.
